File: backend/event-service/src/main/java/lk/nibm/eventservice/service/EventService.java
package lk.nibm.eventservice.service;

import lk.nibm.eventservice.model.Event;
import lk.nibm.eventservice.repository.EventRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class EventService {

    private final EventRepository eventRepository;

    @Autowired
    public EventService(EventRepository eventRepository) {
        this.eventRepository = eventRepository;
    }

    // -----------------------------------------------------------
    // Standard CRUD Methods (Create, Read, Update, Delete)
    // -----------------------------------------------------------

    /**
     * Retrieve all events.
     * Standard CRUD Read (findAll).
     *
     * @return list of all events.
     */
    public List<Event> getEvents() {
        return eventRepository.findAll();
    }

    /**
     * Retrieve a specific event by ID.
     * Standard CRUD Read (findById).
     *
     * @param id the ID of the event.
     * @return the event if found, null otherwise.
     */
    public Event getEventById(int id) {
        return eventRepository.findById(id).orElse(null);
    }

    /**
     * Create a new event.
     * Standard CRUD Create.
     *
     * @param event the event to create.
     * @return the saved event.
     */
    public Event createEvent(Event event) {
        return eventRepository.save(event);
    }

    /**
     * Update an existing event by ID.
     * Standard CRUD Update.
     *
     * @param id the ID of the event to update.
     * @param event the updated event object.
     * @return the updated event if it exists, null otherwise.
     */
    public Event updateEvent(int id, Event event) {
        if (!eventRepository.existsById(id)) {
            return null; // Handle the case where the event does not exist
        }
        event.setId(id); // Ensure the event has the correct ID
        return eventRepository.save(event);
    }

    /**
     * Delete an event by ID.
     * Standard CRUD Delete.
     *
     * @param id the ID of the event to delete.
     * @return true if the event was deleted, false if it doesn't exist.
     */
    public boolean deleteEventById(int id) {
        if (!eventRepository.existsById(id)) {
            return false; // Handle the case where the event does not exist
        }
        eventRepository.deleteById(id);
        return true;
    }

    // -----------------------------------------------------------
    // Custom Methods
    // -----------------------------------------------------------

    /**
     * Search for events by title and optionally by category.
     * Custom search method.
     *
     * @param title the title to search for.
     * @param category optional event category to filter by.
     * @return list of matching events.
     */
    public List<Event> searchEvents(String title, Event.EventCategory category) {
        if (category == null) {
            return eventRepository.findByTitleContainingIgnoreCase(title);
        } else {
            return eventRepository.findByTitleContainingIgnoreCaseAndCategory(title, category);
        }
    }

    /**
     * Retrieve all available event categories.
     * Custom method.
     *
     * @return list of category names.
     */
    public List<String> getCategories() {
        return Arrays.stream(Event.EventCategory.values())
                .map(Enum::name)
                .collect(Collectors.toList());
    }

    // -----------------------------------------------------------
    // Ticket-Related Methods
    // -----------------------------------------------------------

    /**
     * Retrieve available tickets for an event by its ID.
     *
     * @param eventId the ID of the event.
     * @return number of available tickets, or 0 if event is not found.
     */
    public int getAvailableTickets(int eventId) {
        Event event = eventRepository.findById(eventId).orElse(null);
        return (event != null) ? event.getAvailableTickets() : 0;
    }

    /**
     * Set the available tickets for an event.
     *
     * @param eventId the ID of the event.
     * @param availableTickets the number of available tickets to set.
     */
    public void setAvailableTickets(int eventId, int availableTickets) {
        Event event = eventRepository.findById(eventId).orElse(null);
        if (event != null) {
            event.setAvailableTickets(availableTickets);
            eventRepository.save(event);
        }
    }

    /**
     * Retrieve the unit price (ticket price) for an event by its ID.
     *
     * @param eventId the ID of the event.
     * @return the ticket price, or null if event is not found.
     */
    public BigDecimal getUnitPrice(int eventId) {
        Event event = eventRepository.findById(eventId).orElse(null);
        return (event != null) ? event.getTicketPrice() : null;
    }

    /**
     * Set the unit price (ticket price) for an event.
     *
     * @param eventId the ID of the event.
     * @param unitPrice the new ticket price to set.
     */
    public void setUnitPrice(int eventId, BigDecimal unitPrice) {
        Event event = eventRepository.findById(eventId).orElse(null);
        if (event != null) {
            event.setTicketPrice(unitPrice);
            eventRepository.save(event);
        }
    }
}

File: frontend/src/lib/services/eventService.js
import axios from 'axios';

const EVENT_API_URL = 'http://localhost:8082/api/events/'; // Base URL for event-related endpoints

// Create an axios instance with default configurations
const apiClient = axios.create({
	baseURL: EVENT_API_URL,
	headers: {
		'Content-Type': 'application/json'
	}
});

// Add an interceptor to include the Authorization header
apiClient.interceptors.request.use(
	(config) => {
		const token = localStorage.getItem('token');
		if (token) {
			config.headers.Authorization = `Bearer ${token}`;
		}
		return config;
	},
	(error) => {
		return Promise.reject(error);
	}
);

export const eventService = {
	// Retrieve all events
	getAllEvents: () => apiClient.get('/all').then((response) => response.data),

	// Retrieve an event by ID
	getEventById: (id) => apiClient.get(`/${id}`).then((response) => response.data),

	// Search for events by title and/or category
	searchEvents: (title, category) =>
		apiClient.get('/search', { params: { title, category } }).then((response) => response.data),

	// Create a new event
	createEvent: (eventData, organizerId) =>
		apiClient.post('/', { ...eventData, organizerId }).then((response) => response.data),

	// Update an event by ID
	updateEvent: (id, eventData) =>
		apiClient.put(`/${id}`, eventData).then((response) => response.data),

	// Delete an event by ID
	deleteEvent: (id) => apiClient.delete(`/${id}`),

	// Retrieve all available event categories
	getCategories: () => apiClient.get('/categories').then((response) => response.data),

	// Retrieve the number of available tickets for an event by ID
	getAvailableTickets: (id) =>
		apiClient.get(`/tickets/available/${id}`).then((response) => response.data),

	// Set the number of available tickets for an event by ID
	setAvailableTickets: (id, availableTickets) =>
		apiClient.put(`/tickets/available/${id}`, null, { params: { availableTickets } }),

	// Retrieve the ticket price (unit price) for an event by ID
	getUnitPrice: (id) => apiClient.get(`/tickets/price/${id}`).then((response) => response.data),

	// Set the ticket price (unit price) for an event by ID
	setUnitPrice: (id, unitPrice) =>
		apiClient.put(`/tickets/price/${id}`, null, { params: { unitPrice } })
};
